src/lib.rs

```diff
@@ -0,0 +1,550 @@
+/// - tokio, all other flavors - spawn blocking
+/// 
+/// You can override these defaults by initializing a strategy:
+/// - [`initialize_current_context_strategy`],
+/// - [`initialize_spawn_blocking_strategy`]
+/// - [`initialize_block_in_place_strategy`]
+/// - [`initialize_secondary_tokio_runtime_strategy`]
+/// - [`initialize_secondary_tokio_runtime_strategy_and_config`]
+/// 
+/// # Example 
+/// 
+/// ```
+/// # async fn run() {
+/// use compute_heavy_future_executor::spawn_compute_heavy_future;
+/// 
+/// let future = async { 
+///     std::thread::sleep(std::time::Duration::from_millis(50));
+///     5
+///  }; 
+/// 
+/// let res = spawn_compute_heavy_future(future).await.unwrap();
+/// assert_eq!(res, 5);
+/// # }
+/// 
+/// ```
+/// 
+pub async fn spawn_compute_heavy_future<F, R>(fut: F) -> Result<R, Error>
+where
+    F: Future<Output = R> + Send + 'static,
+    R: Send + 'static,
+{
+    let executor = COMPUTE_HEAVY_FUTURE_EXECUTOR_STRATEGY
+        .get_or_init(|| {
+                #[cfg(compute_heavy_executor_tokio)]
+                match tokio::runtime::Handle::current().runtime_flavor() {
+                    tokio::runtime::RuntimeFlavor::MultiThread => {
+                        log::info!("spawn_compute_heavy_future called without setting an explicit executor strategy, \
+                        setting to BlockInPlace due to tokio multithreaded runtime context");
+                        return ExecutorStrategy::BlockInPlace(BlockInPlaceExecutor {})
+                    },
+                    _ => {
+                        log::info!("spawn_compute_heavy_future called without setting an explicit executor strategy, \
+                        setting to SpawnBlocking");
+                        return ExecutorStrategy::SpawnBlocking(SpawnBlockingExecutor {})
+                    },
+                };
+
+                #[allow(unreachable_code)]
+                {
+                    log::warn!("spawn_compute_heavy_future called without setting an explicit executor strategy, setting to \
+                    current context due to no `cfg(compute_heavy_executor_tokio)`. This is a non-op and probably not what you want.");
+                    ExecutorStrategy::CurrentContext(CurrentContextExecutor {})    
+                }
+        });
+    match executor {
+        ExecutorStrategy::CurrentContext(executor) => executor.execute(fut).await,
+        ExecutorStrategy::CustomExecutor(executor) => executor.execute(fut).await,
+        #[cfg(compute_heavy_executor_tokio)]
+        ExecutorStrategy::BlockInPlace(executor) => executor.execute(fut).await,
+        #[cfg(compute_heavy_executor_tokio)]
+        ExecutorStrategy::SpawnBlocking(executor) => executor.execute(fut).await,
+        #[cfg(compute_heavy_executor_tokio)]
+        ExecutorStrategy::SecondaryTokioRuntime(executor) => executor.execute(fut).await,
+    }
+}
+
+// Use nextest to run these, otherwise there will be race conditions against the strategy oncecell
```

Comment:
Use doctests to ensure separate processes?

Reply:
Oh, is that conventional? I can poke around at that.

I figured I can run them however i want (with nextest) in ci/cd anyway. But probably better to make it 'just work' locally.

Reply:
We discussed, I'll just use integration tests.